snuba/datasets/storages/metrics.py

```diff
@@ -102,7 +105,9 @@
     ),
     query_processors=[ArrayJoinKeyValueOptimizer("tags"), TableRateLimit()],
     stream_loader=build_kafka_stream_loader_from_settings(
-        SetsAggregateProcessor(), default_topic=Topic.METRICS,
+        SetsAggregateProcessor(),
+        default_topic=Topic.METRICS,
+        dead_letter_queue_policy=IgnoreInvalidMessagePolicy(),
```

Comment:
In order for the IgnoreInvalidMessagePolicy to be useful, when it drops a message and records the metric it needs to be able to provide context. Now we would not know which consumer received a bad message. 

Though there is a larger issue here: https://github.com/getsentry/arroyo/blob/d05860814366ff7d37ed90346385439de600c7bc/arroyo/processing/strategies/dead_letter_queue/policies/ignore.py#L10
Metrics will be initialized at the initialization of the class. This is a problem since now you are creating the policy at the initialization of the storage, which likely happens before this line 
https://github.com/getsentry/snuba/blob/master/snuba/cli/consumer.py#L125-L130
Which is when we set up the metrics. 
If your policy is initialized before that line (as it will) your policy will be initialized with a dummy metrics backend and never record any metric.

In practice the issue is that the data model should not reference arroyo components that rely on metrics initialization. @lynnagara  we may want to revisit the design decision of initializing metrics and have them as a singleton for arroyo. This can bite us back in these scenarios.

You have two ways to solve this:
- do not initialize the policy here. Provide instead a closure that initialize the policy. This closure will be called by KafkaConsumerStrategyFactory when the strategy is created.
- make the `IgnoreInvalidMessagePolicy` load the metrics each time it needs them, but this is going to be a problem as it creates an implicit coupling between the way arroyo works and the way snuba works. The two should exist independently from each other or, at best , Snuba should adapt to arroyo as Snuba is a client of the library, not vice versa.

Reply:
> we would not know which consumer received a bad message.

If the closure solution to the larger problem works, does this still apply? It seems the metrics will have the consumer group and storage key tags.

Reply:
How do you plan to support the policy that needs a Kafka producer to be provided to the constructor ?

Reply:
Wouldn't this work?

```python3
def produce_policy_closure() -> DeadLetterQueuePolicy:
    return ProduceInvalidMessagePolicy(
        KafkaProducer(build_kafka_producer_configuration(dead_letter_topic)),
        dead_letter_topic,
    )
```